lib/explorer/data_frame.ex

```diff
@@ -2800,77 +2800,105 @@ defmodule Explorer.DataFrame do
 
     result = fun.(ldf)
 
-    column_pairs =
-      to_column_pairs(df, result, fn value ->
-        case value do
-          %Series{data: %LazySeries{}} = lazy_series ->
-            lazy_series
+    column_pairs = to_column_pairs(df, result, &value!/1)
 
-          %Series{data: _other} ->
-            raise ArgumentError,
-                  "expecting a lazy series. Consider using `Explorer.DataFrame.put/3` " <>
-                    "to add eager series to your dataframe."
+    new_dtypes =
+      for {column_name, series} <- column_pairs, into: %{} do
+        {column_name, series.dtype}
+      end
 
-          list when is_list(list) ->
-            raise ArgumentError,
-                  "expecting a lazy series or scalar value, but instead got a list. " <>
-                    "consider using `Explorer.Series.from_list/2` to create a `Series`, " <>
-                    "and then `Explorer.DataFrame.put/3` to add the series to your dataframe."
+    mut_names = Enum.map(column_pairs, &elem(&1, 0))
+    new_names = Enum.uniq(df.names ++ mut_names)
 
-          nil ->
-            lazy_s = LazySeries.new(:lazy, [nil], :null)
-            Explorer.Backend.Series.new(lazy_s, :null)
+    df_out = %{df | names: new_names, dtypes: Map.merge(df.dtypes, new_dtypes)}
 
-          number when is_number(number) ->
-            dtype = if is_integer(number), do: {:s, 64}, else: {:f, 64}
-            lazy_s = LazySeries.new(:lazy, [number], dtype)
+    column_pairs = for {name, %Series{data: lazy_series}} <- column_pairs, do: {name, lazy_series}
 
-            Explorer.Backend.Series.new(lazy_s, dtype)
+    Shared.apply_impl(df, :mutate_with, [df_out, column_pairs])
+  end
 
-          string when is_binary(string) ->
-            lazy_s = LazySeries.new(:lazy, [string], :string)
+  defp value!(%Series{data: %LazySeries{}} = lazy_series) do
+    lazy_series
+  end
 
-            Explorer.Backend.Series.new(lazy_s, :string)
+  defp value!(%Series{data: _other}) do
+    raise ArgumentError,
+          "expecting a lazy series. Consider using `Explorer.DataFrame.put/3` " <>
+            "to add eager series to your dataframe."
+  end
 
-          boolean when is_boolean(boolean) ->
-            lazy_s = LazySeries.new(:lazy, [boolean], :boolean)
+  defp value!(list) when is_list(list) do
```

Comment:
lists can be converted to map, name will be inferred if it is a plain column without op. For everything else, name should be specified i.e {name, Series}. 

```elixir
DF.mutate(df, c: [a, b])
#Explorer.DataFrame<
  Polars[3 x 3]
  a s64 [1, nil, 3]
  b string ["a", "b", nil]
  c struct[2] [
    %{"a" => 1, "b" => "a"},
    %{"a" => nil, "b" => "b"},
    %{"a" => 3, "b" => nil}
  ]
>

DF.mutate(df, c: [a, b, a_nil: is_nil(a)])
#Explorer.DataFrame<
  Polars[3 x 3]
  a s64 [1, nil, 3]
  b string ["a", "b", nil]
  c struct[3] [
    %{"a" => 1, "a_nil" => false, "b" => "a"},
    %{"a" => nil, "a_nil" => true, "b" => "b"},
    %{"a" => 3, "a_nil" => false, ...}
  ]
>
```

Reply:
can revert it if you want.

Reply:
Yes, let's please revert it, as we don't support this anywhere in Elixir and we can use the list syntax for creating literal lists as well.